tests/Microsoft.Identity.Test.Unit/ManagedIdentityTests/ManagedIdentityTests.cs

```diff
@@ -480,5 +483,168 @@ public async Task ManagedIdentityCacheTestAsync()
                 appTokenCacheRecoder.AssertAccessCounts(1, 1);
             }
         }
+
+        [DataTestMethod]
+        [DataRow(0)]
+        [DataRow(1)]
+        [DataRow(2)]
+        [DataRow(3)]
+        public async Task ManagedIdentityRefreshOnIsCalculatedAsync(int expiresInHours)
+        {
+            AuthenticationResult result;
+
+            using (new EnvVariableContext())
+            using (var httpManager = new MockHttpManager())
+            {
+                SetEnvironmentVariables(ManagedIdentitySource.AppService, AppServiceEndpoint);
+
+                var mi = ManagedIdentityApplicationBuilder.Create()
+                    .WithExperimentalFeatures()
+                    .WithHttpManager(httpManager)
+                    .Build();
+
+                httpManager.AddManagedIdentityMockHandler(
+                    AppServiceEndpoint,
+                    Resource,
+                    MockHelpers.GetMsiSuccessfulResponse(expiresInHours),
+                    ManagedIdentitySource.AppService);
+
+                AcquireTokenForManagedIdentityParameterBuilder builder = mi.AcquireTokenForManagedIdentity(Resource);
+
+                if (expiresInHours == 0)
```

Comment:
nit: this can be moved to a separate test method, something like InvalidRefreshOn_Throws